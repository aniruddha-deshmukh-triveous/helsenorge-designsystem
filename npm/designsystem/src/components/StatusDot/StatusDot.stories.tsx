import React from 'react';
import { withKnobs, text, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import StatusDot, { StatusDotVariant } from './StatusDot';
import { withA11y } from '@storybook/addon-a11y';

const stories = storiesOf('StatusDot', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{ width: '20rem' }}>
    <StatusDot variant={select('Variant', StatusDotVariant, StatusDotVariant.info)} text={text('Text', 'StatusDot Text')} />
  </div>
));

stories.add('All types', () => (
  <div style={{ width: '20rem' }}>
    <p>{'info'}</p>
    <StatusDot variant={StatusDotVariant.info} text="Info" />
    <p>{'warning'}</p>
    <StatusDot variant={StatusDotVariant.warning} text="Warning" />
    <p>{'alert'}</p>
    <StatusDot variant={StatusDotVariant.alert} text="Alert" />
    <p>{'cancelled'}</p>
    <StatusDot variant={StatusDotVariant.cancelled} text="Cancelled" />
    <p>{'active'}</p>
    <StatusDot variant={StatusDotVariant.active} text="Active" />
    <p>{'transparent'}</p>
    <StatusDot variant={StatusDotVariant.transparent} text="Transparent" />
    <p>{'recurring'}</p>
    <StatusDot variant={StatusDotVariant.recurring} text="Recurring" />
    <p>{'group'}</p>
    <StatusDot variant={StatusDotVariant.group} text="Group" />
    <p>{'info no text'}</p>
    <StatusDot variant={StatusDotVariant.info} text="" />
  </div>
));
