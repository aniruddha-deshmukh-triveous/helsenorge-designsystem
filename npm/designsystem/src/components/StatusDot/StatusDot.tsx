import React from 'react';

import classNames from 'classnames';

import { IconSize } from '../..';
import { AnalyticsId } from '../../constants';
import { getColor } from '../../theme/currys';
import Attachment from '../Icons/Attachment';
import Group from '../Icons/Group';
import { Icon } from '../Icons/Icon';
import NoAccess from '../Icons/NoAccess';
import Undo from '../Icons/Undo';

import styles from './styles.module.scss';

export enum StatusDotVariant {
  info = 'info',
  warning = 'warning',
  alert = 'alert',
  cancelled = 'cancelled',
  active = 'active',
  transparent = 'transparent',
  recurring = 'recurring',
  group = 'group',
  noaccess = 'noaccess',
  attachment = 'attachment',
}

export interface StatusDotIconProps {
  variant?: keyof typeof StatusDotVariant;
}

const StatusDotIcon: React.FC<StatusDotIconProps> = ({ variant }) => {
  if (variant === StatusDotVariant.recurring) {
    return <Icon size={IconSize.XXSmall} svgIcon={Undo} />;
  } else if (variant === StatusDotVariant.group) {
    return <Icon size={IconSize.XXSmall} svgIcon={Group} />;
  } else if (variant === StatusDotVariant.noaccess) {
    return <Icon size={IconSize.XXSmall} svgIcon={NoAccess} color={getColor('cherry', 600)} />;
  } else if (variant === StatusDotVariant.attachment) {
    return <Icon size={IconSize.XXSmall} svgIcon={Attachment} />;
  }

  return null;
};

export interface StatusDotProps {
  /** Visual variants for the statusdot */
  variant?: keyof typeof StatusDotVariant;
  /** Text placed to the right of the statusdot */
  text: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

const StatusDot: React.FC<StatusDotProps> = props => {
  const { variant = StatusDotVariant.info, text, className, testId } = props;

  const hasIcon =
    variant === StatusDotVariant.recurring ||
    variant === StatusDotVariant.group ||
    variant === StatusDotVariant.noaccess ||
    variant === StatusDotVariant.attachment;

  const isCancelled = variant === StatusDotVariant.cancelled;

  const statusDotClasses = classNames(styles['statusdot'], isCancelled && styles['statusdot--cancelled'], className);
  const dotClasses = classNames(styles['statusdot__dot'], hasIcon ? styles[`statusdot__dot--icon`] : styles[`statusdot__dot--${variant}`]);
  const labelClasses = classNames(styles['statusdot__label'], hasIcon && styles[`statusdot__label--icon`]);

  return (
    <div className={statusDotClasses} data-testid={testId} data-analyticsid={AnalyticsId.StatusDot}>
      <span className={dotClasses}>
        <StatusDotIcon variant={variant} />
      </span>
      <span className={labelClasses}>{text}</span>
    </div>
  );
};

export default StatusDot;
