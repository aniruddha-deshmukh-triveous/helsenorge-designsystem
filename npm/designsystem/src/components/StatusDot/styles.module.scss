@import '../../scss/palette';
@import '../../scss/spacers';
@import '../../scss/font-settings';
@import '../../scss/screen-reader';

.statusdot {
  position: relative;
  display: inline-flex;
  align-items: center;
  font-size: $font-size-sm;
  font-weight: 400;
  line-height: $lineheight-size-sm;

  &--cancelled {
    color: $cherry600;
  }

  &__dot {
    top: getSpacer(3xs);
    margin-right: getSpacer(2xs);
    border-radius: 10rem;

    &:not(#{&}--icon) {
      width: 0.937rem;
      height: 0.937rem;
    }

    &--info {
      background: $kiwi900;
    }

    &--warning {
      background: $banana400;
    }

    &--alert,
    &--cancelled {
      background: $cherry500;
    }

    &--active {
      background: $blueberry500;
    }

    &--transparent {
      border: 1px solid $neutral500;
    }

    &--icon {
      position: absolute;
      margin: getSpacer(4xs) * -1 0 0 getSpacer(3xs) * -1;
    }
  }

  &__label {
    &--icon {
      margin-left: getSpacer(m);
    }
  }
}
