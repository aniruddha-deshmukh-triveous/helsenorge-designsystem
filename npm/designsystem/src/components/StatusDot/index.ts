import StatusDot from './StatusDot';
export * from './StatusDot';
export default StatusDot;
