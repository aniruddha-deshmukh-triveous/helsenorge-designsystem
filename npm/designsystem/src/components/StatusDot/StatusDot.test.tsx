import * as React from 'react';
import StatusDot from '.';
import { render, screen } from '@testing-library/react';

describe('Gitt at StatusDot rendres ', () => {
  describe('Når komponenten får en tekst', () => {
    it('Så skal teskten vises til innbygger', () => {
      render(<StatusDot variant="warning" text="Eksempeltekst" />);
      expect(screen.queryByText('Eksempeltekst')).toBeTruthy();
    });
  });
  describe('Når StatusDot har type="recurring".', () => {
    it('Så skal komponenten ha en SVG-fil.', () => {
      render(<StatusDot variant="recurring" text="Test Text" />);
      expect(screen.queryByText('Test Text').parentElement.firstElementChild.className).toBe('statusdot__dot statusdot__dot--icon');
      expect(screen.queryByText('Test Text').parentElement.lastElementChild.className).toBe('statusdot__label statusdot__label--icon');

      render(<StatusDot variant="group" text="Test Text2" />);
      expect(screen.queryByText('Test Text2').parentElement.firstElementChild.className).toBe('statusdot__dot statusdot__dot--icon');
      expect(screen.queryByText('Test Text2').parentElement.lastElementChild.className).toBe('statusdot__label statusdot__label--icon');

      render(<StatusDot variant="noaccess" text="Test Text3" />);
      expect(screen.queryByText('Test Text3').parentElement.firstElementChild.className).toBe('statusdot__dot statusdot__dot--icon');
      expect(screen.queryByText('Test Text3').parentElement.lastElementChild.className).toBe('statusdot__label statusdot__label--icon');
    });
  });
  describe('Når StatusDot har en av de andre typene', () => {
    it('Så skal komponenten ikke ha en tilhørende SVG-fil og samsvarende klassenavn', () => {
      render(<StatusDot variant="info" text="text1" />);
      render(<StatusDot variant="warning" text="text2" />);
      render(<StatusDot variant="alert" text="text3" />);
      render(<StatusDot variant="transparent" text="text4" />);

      const dot1 = screen.queryByText('text1').parentElement.firstElementChild;
      const dot2 = screen.queryByText('text2').parentElement.firstElementChild;
      const dot3 = screen.queryByText('text3').parentElement.firstElementChild;
      const dot4 = screen.queryByText('text4').parentElement.firstElementChild;

      expect(dot1.className).toBe('statusdot__dot statusdot__dot--info');
      expect(dot2.className).toBe('statusdot__dot statusdot__dot--warning');
      expect(dot3.className).toBe('statusdot__dot statusdot__dot--alert');
      expect(dot4.className).toBe('statusdot__dot statusdot__dot--transparent');
    });
  });
});
