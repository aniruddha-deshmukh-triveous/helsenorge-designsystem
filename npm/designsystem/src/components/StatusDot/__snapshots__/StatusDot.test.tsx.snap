// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at StatusDot rendres  Når StatusDot har type="attachment". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <div
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M12 4.586c-1.798 0-3.27 1.384-3.27 3.1v11.23h1.005V7.686c0-1.148 1.009-2.095 2.265-2.095 1.257 0 2.265.947 2.265 2.094V17.6c0 .5-.355.81-.868.81s-.861-.341-.861-.81V9.467H11.53v8.132c0 1.14.833 1.815 1.867 1.815s1.874-.83 1.874-1.815V7.685c0-1.715-1.473-3.1-3.271-3.1Z"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </div>
</div>
`;

exports[`Gitt at StatusDot rendres  Når StatusDot har type="group". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <div
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M8.812 5.18h-.274a4.639 4.639 0 0 0-4.644 4.65v2.276a4.639 4.639 0 0 0 4.644 4.65h.274a4.64 4.64 0 0 0 4.644-4.65V9.83a4.64 4.64 0 0 0-4.644-4.65m4.07 14.526c-.076-.05-.15-.076-.226-.126a8.236 8.236 0 0 1-3.844.926h-.274a8.113 8.113 0 0 1-3.796-.926A8.612 8.612 0 0 0 0 27.306v2.1h17.376v-2.1c0-3.3-1.798-6.15-4.494-7.6m12.356-5.55a8.235 8.235 0 0 1-3.844.924h-.274a8.104 8.104 0 0 1-3.794-.924 9.173 9.173 0 0 0-2.772 2.198c4.044 2.152 6.566 6.3 6.566 10.952v2.1h8.838V21.88c0-3.4-1.898-6.3-4.72-7.724m.864-9.504v2.282a4.646 4.646 0 0 1-4.644 4.65h-.274a4.648 4.648 0 0 1-4.646-4.65V4.652c0-2.57 2.08-4.652 4.646-4.652h.274a4.647 4.647 0 0 1 4.644 4.652"
          fill-rule="evenodd"
          transform="translate(8, 8.2966)"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </div>
</div>
`;

exports[`Gitt at StatusDot rendres  Når StatusDot har type="noaccess". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <div
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="#B62E1C"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M13.4 26.1h21.2v-4.2H13.4v4.2ZM24 9C15.728 9 9 15.728 9 24s6.728 15 15 15c8.27 0 15-6.728 15-15S32.27 9 24 9Z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </div>
</div>
`;

exports[`Gitt at StatusDot rendres  Når StatusDot har type="recurring". Så skal komponenten ha en SVG-fil. 1`] = `
<div>
  <div
    class="statusdot"
    data-analyticsid="status-dot"
    data-testid="statusdot"
  >
    <span
      class="statusdot__dot statusdot__dot--icon"
    >
      <svg
        aria-hidden="true"
        class="hnds-style-icon"
        data-analyticsid="icon"
        fill="black"
        focusable="false"
        height="24"
        role="presentation"
        style="min-width: 24px; min-height: 24px;"
        viewBox="0 0 48 48"
        width="24"
      >
        <path
          d="M38.726 24.194c0 8.196-6.668 14.864-14.864 14.864-7.028 0-13.152-4.992-14.56-11.87l2.744-.56c1.14 5.578 6.11 9.628 11.816 9.628 6.652 0 12.064-5.41 12.064-12.062 0-6.652-5.412-12.062-12.064-12.062-3.1 0-6.05 1.218-8.272 3.312h5.248v2.802h-9.516l-.004-9.89h2.802v4.624a14.893 14.893 0 0 1 9.742-3.648c8.196 0 14.864 6.666 14.864 14.862"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="statusdot__label statusdot__label--icon"
    >
      Eksempeltekst
    </span>
  </div>
</div>
`;
